Increase support for IronSide SE UICR features

SebastianBoe opened the pull request:
Increase support for IronSide SE UICR features

SebastianBoe committed:
soc: nordic: uicr: Add support for UICR.WDTSTART

Add support for UICR.WDTSTART.

UICR.WDTSTART configures the automatic start of a local watchdog timer
before the application core is booted. This provides early system
protection ensuring that the system can recover from early boot
failures.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
soc: nordic: uicr: Add support for UICR.SECONDARY.TRIGGER

Add support for UICR.SECONDARY.TRIGGER configuration, which enables
automatic booting of secondary firmware based on specific reset reasons.

This introduces Kconfig options for configuring:
- UICR.SECONDARY.TRIGGER.ENABLE - Enable/disable automatic triggers
- UICR.SECONDARY.TRIGGER.RESETREAS - Bitmask of reset reasons that
  trigger secondary firmware boot

Individual Kconfig options are provided for each reset reason:
- APPLICATIONWDT0/1 - Application core watchdog timeouts
- APPLICATIONLOCKUP - Application core CPU lockup
- RADIOCOREWDT0/1 - Radio core watchdog timeouts
- RADIOCORELOCKUP - Radio core CPU lockup

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
soc: nordic: uicr: Add support for UICR.SECONDARY.PROTECTEDMEM

Add support for UICR.SECONDARY.PROTECTEDMEM configuration, which enables
configuration of the protected memory region for secondary firmware.

This introduces Kconfig options for configuring:
- GEN_UICR_SECONDARY_PROTECTEDMEM - Enable/disable protected memory
  for secondary firmware
- GEN_UICR_SECONDARY_PROTECTEDMEM_SIZE_BYTES - Size of the protected
  memory region in bytes

The implementation validates that the configured size is divisible by
4096 bytes (4 KiB) as required by the hardware, and converts it to
4 KiB units when writing to UICR.SECONDARY.PROTECTEDMEM.SIZE4KB.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
soc: nordic: uicr: Add support for UICR.LOCK

Add support for UICR.LOCK configuration, which locks the entire UICR
configuration in NVR0 to prevent unauthorized modifications.

This introduces a Kconfig option GEN_UICR_LOCK that enables locking
of the UICR. Once locked, the UICR can only be modified by performing
an ERASEALL operation.

This is a critical security feature for production devices, typically
enabled alongside UICR.APPROTECT, UICR.PROTECTEDMEM, and
UICR.ERASEPROTECT to establish a complete device protection scheme.

When enabled, the gen_uicr.py script sets UICR.LOCK to 0xFFFFFFFF,
which configures the NVR0 page as read-only and enforces integrity
checks on the UICR content.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
soc: nordic: uicr: Add support for UICR.ERASEPROTECT

Add support for UICR.ERASEPROTECT configuration, which blocks ERASEALL
operations to prevent bulk erasure of protected memory.

This introduces a Kconfig option GEN_UICR_ERASEPROTECT that enables
blocking of ERASEALL operations on NVR0, preserving UICR settings even
if an attacker attempts a full-chip erase.

This is a critical security feature for production devices. When enabled
together with UICR.LOCK, it becomes impossible to modify the UICR in
any way, establishing a permanent device protection scheme. Due to this
irreversibility, it should only be enabled during the final stages of
production.

When enabled, the gen_uicr.py script sets UICR.ERASEPROTECT to
0xFFFFFFFF, which prevents the ERASEALL command from affecting the
NVR0 page.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
soc: nordic: uicr: Add support for UICR.APPROTECT

Add support for UICR.APPROTECT configuration, which controls debugger
and access-port permissions through the TAMPC peripheral.

This introduces three Kconfig options that allow independent control
over access port protection for different processor domains:

- GEN_UICR_APPROTECT_APPLICATION_PROTECTED: Controls debug access to
  the application domain processor
- GEN_UICR_APPROTECT_RADIOCORE_PROTECTED: Controls debug access to
  the radio core processor
- GEN_UICR_APPROTECT_CORESIGHT_PROTECTED: Controls access to the
  CoreSight debug infrastructure

When enabled, each option sets the corresponding UICR.APPROTECT
register to PROTECTED (0xFFFFFFFF), which disables debug access for
that domain. When disabled, the registers remain at their erased value
(UNPROTECTED), allowing full debug access.

This feature is critical for production devices where debug access must
be restricted to prevent unauthorized access to sensitive code and data.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

hakonfam commented on scripts/ci/check_compliance.py:
```diff
@@ -1266,12 +1266,18 @@ def check_no_undef_outside_kconfig(self, kconf):
+        "GEN_UICR_SECURESTORAGE",
+        "GEN_UICR_WDTSTART",
+        "GEN_UICR_WDTSTART_CRV",
+        "GEN_UICR_WDTSTART_INSTANCE_CODE", # Used in specialized build tool, not part of main Kconfig
```
nit: is the purpose here to only have the comment on the last entry? If so the top entry still has the comment.

SebastianBoe replied:
The problem with having at the beginning and end is that alphabetical sorting is required so the comment would have to move from time to time, and this pollutes commits with unrelated changes.

hakonfam commented on soc/nordic/common/uicr/gen_uicr.py:
```diff
@@ -440,6 +440,36 @@ def main() -> None:
+        type=lambda s: int(s, 0),
+        help="Watchdog timer instance code (0xBD2328A8 for WDT0, 0x1730C77F for WDT1)",
+    )
+    parser.add_argument(
```
Should we use a sub-parer for the fields that are only relevant if ``--wdtstart`` is enabled?

I see that we handle this in kconfig, but the script would be more robust for isolated use with that change IMO.

hakonfam commented on soc/nordic/common/uicr/gen_uicr/Kconfig:
```diff
@@ -46,6 +46,91 @@ config GEN_UICR_PROTECTEDMEM_SIZE_BYTES
+	help
+	  When enabled, the UICR generator will configure the
+	  watchdog timer to start automatically before the
+	  application core is booted.
```
Should we mention that its the local watchdog timer instance as opposed to the global one?

hakonfam commented on soc/nordic/common/uicr/gen_uicr/Kconfig:
```diff
@@ -32,6 +32,15 @@ config GEN_UICR_SECURESTORAGE
+	bool "Enable UICR.LOCK"
+	help
+	  When enabled, the UICR generator will lock the entire UICR
+	  configuration in NVR0. Once locked, the UICR can only be modified
```
Doesn't it lock all NVR0? Should we specify that its the NVR0 in MRAM10? Which includes BICR.

hakonfam reviewed (changes requested):
We need a mechanism stopping users from enabling UICR.LOCK and UICR.ERASEPROTECT at the same time by accident.

This configuration does not give much value as one would typically complete the device lock down in run time in the final stages of production.

hakonfam commented on soc/nordic/common/uicr/Kconfig:
```diff
@@ -18,3 +18,10 @@ config NRF_PERIPHCONF_GENERATE_ENTRIES
 	  device configuration in the devicetree.
+
+config IS_IRONSIDE_SE_SECONDARY_IMAGE
+	bool "Ironside SE secondary image indicator (informative only, do not change)"
```
Yeah, then we need the prompt.

hakonfam reviewed (changes requested):
NACK since IMO we need further hardening against creating "bricking" UICRs.

SebastianBoe committed:
soc: nordic: uicr: Add safety flag for permanent device transition

Add --permit-permanently-transitioning-device-to-deployed safety flag
to gen_uicr.py, required when enabling both UICR.LOCK and
UICR.ERASEPROTECT together. This prevents accidental permanent locking
of devices since this combination makes the configuration irreversible.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

karstenkoenig reviewed (dismissed):
Only cosmetic feedback
